src/signTx.c

```diff
@@ -0,0 +1,648 @@
+	case POLICY_PROMPT_BEFORE_RESPONSE: {
+		ctx->ui_step = HANDLE_INIT_STEP_CONFIRM;
+		break;
+	}
+	case POLICY_ALLOW: {
+		ctx->ui_step = HANDLE_INIT_STEP_RESPOND;
+		break;
+	}
+	default:
+		THROW(ERR_NOT_IMPLEMENTED);
+	}
+	signTx_handleInit_ui_runStep();
+}
+
+// Simple for now
+static void signTx_handleInit_ui_runStep()
+{
+	ui_callback_fn_t* this_fn = signTx_handleInit_ui_runStep;
+	int nextStep = HANDLE_INIT_STEP_INVALID;
+
+	switch(ctx->ui_step) {
+	case HANDLE_INIT_STEP_CONFIRM: {
+		TRACE();
+		ui_displayConfirm(
+		        "Start new",
+		        "transaction?",
+		        this_fn,
+		        respond_with_user_reject
+		);
+		nextStep = HANDLE_INIT_STEP_RESPOND;
+		break;
+	}
+	case HANDLE_INIT_STEP_RESPOND: {
+		TRACE();
+		// advance state to inputs
+		txHashBuilder_enterInputs(&ctx->txHashBuilder);
+		ctx->stage = SIGN_STAGE_INPUTS;
+
+		// respond
+		io_send_buf(SUCCESS, NULL, 0);
+		ui_displayBusy(); // needs to happen after I/O
+		nextStep = HANDLE_INIT_STEP_INVALID;
+		break;
+	}
+	default:
+		ASSERT(false);
+	}
+	ctx->ui_step = nextStep;
+}
+
+
+static void amountSum_incrementBy(uint64_t* sumPtr, uint64_t amount)
+{
+	ASSERT(*sumPtr <= LOVELACE_MAX_SUPPLY);
+	ASSERT(*sumPtr <= LOVELACE_MAX_SUPPLY);
+	if (*sumPtr + amount > LOVELACE_MAX_SUPPLY) {
+		THROW(ERR_INVALID_DATA);
+	}
+	*sumPtr += amount;
+}
+
+static void signTx_handleInput_ui_runStep();
+
+enum {
+	HANDLE_INPUT_STEP_RESPOND = 200,
+	HANDLE_INPUT_STEP_INVALID,
+};
+
+static void signTx_handleInputAPDU(uint8_t p2, uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	TRACE();
+	ASSERT(ctx->currentInput < ctx->numInputs);
+	CHECK_STAGE(SIGN_STAGE_INPUTS);
+
+	VALIDATE(p2 == 0, ERR_INVALID_REQUEST_PARAMETERS);
+	VALIDATE(wireDataSize >= 1, ERR_INVALID_DATA);
+	uint8_t inputType = wireDataBuffer[0];
+
+	if (inputType == SIGN_TX_INPUT_TYPE_UTXO) {
+
+		struct {
+			struct {
+				uint8_t txHash[32];
+				uint8_t index[4];
+				uint8_t amount[8];
+			} data;
+			uint8_t hmac[16];
+		}* wireUtxo = (void*) wireDataBuffer + 1;
+
+		VALIDATE(wireDataSize == 1 + SIZEOF(*wireUtxo), ERR_INVALID_DATA);
+
+		if (!attest_isCorrectHmac(
+		            ATTEST_PURPOSE_BIND_UTXO_AMOUNT,
+		            (uint8_t*) &wireUtxo->data, SIZEOF(wireUtxo->data),
+		            wireUtxo->hmac, SIZEOF(wireUtxo->hmac)
+		    )) {
+			THROW(ERR_INVALID_DATA);
+		}
+
+		uint8_t* txHashBuffer = wireUtxo->data.txHash;
+		size_t txHashSize = SIZEOF(wireUtxo->data.txHash);
+		uint32_t parsedIndex = u4be_read(wireUtxo->data.index);
+		uint64_t parsedAmount = u8be_read(wireUtxo->data.amount);
+
+		TRACE("Input amount %d %d", (int)(parsedAmount / 1000000), (int)(parsedAmount % 1000000));
+		amountSum_incrementBy(&ctx->sumAmountInputs, parsedAmount);
+
+		TRACE("Adding input to tx hash");
+		txHashBuilder_addUtxoInput(&ctx->txHashBuilder, txHashBuffer, txHashSize, parsedIndex);
+	} else {
+		// Unknown type
+		THROW(ERR_INVALID_DATA);
+	}
+
+	security_policy_t policy = policyForSignTxInput();
+
+	switch (policy) {
```

Comment:
I'd use if since we currently have no more implementations

Reply:
Normally I would agree but I want to keep it consistent with the rest of the file so that it is easier to reason about/orient in the code. Wouldn't you agree?